java/ql/src/experimental/Security/CWE/CWE-089/MyBatisAnnotationSqlInjectionLib.qll

```diff
@@ -0,0 +1,128 @@
+/**
+ * Provides classes for SQL injection detection in MyBatis annotation.
+ */
+
+import java
+import MyBatisCommonLib
+import semmle.code.xml.MyBatisMapperXML
+import semmle.code.java.dataflow.FlowSources
+import semmle.code.java.frameworks.Properties
+
+/** A sink for MyBatis annotation method call an argument. */
+class MyBatisAnnotationMethodCallAnArgument extends DataFlow::Node {
+  MyBatisAnnotationMethodCallAnArgument() {
+    exists(MyBatisSqlOperationAnnotationMethod msoam, MethodAccess ma | ma.getMethod() = msoam |
+      ma.getAnArgument() = this.asExpr()
+    )
+  }
+}
+
+/** Get the #{...} or ${...} parameters in the Mybatis annotation value. */
+private string getAnMybatiAnnotationSetValue(IbatisSqlOperationAnnotation isoa) {
+  result = isoa.getSqlValue().trim().regexpFind("(#|\\$)\\{[^\\}]*\\}", _, _)
+}
+
+predicate isMybatisAnnotationSqlInjection(DataFlow::Node node, IbatisSqlOperationAnnotation isoa) {
+  // MyBatis uses an annotation method to perform SQL operations. This method has only one parameter and
+  // the parameter is not annotated with `@Param`. Improper use of this parameter has a SQL injection vulnerability.
+  // e.g.
+  //
+  // ```java
+  //    @Select(select id,name from test where name like '%${value}%')
+  //    Test test(String name);
+  // ```
+  exists(MyBatisSqlOperationAnnotationMethod msoam, MethodAccess ma, string res |
+    msoam = ma.getMethod()
+  |
+    msoam.getAnAnnotation() = isoa and
+    res = getAnMybatiAnnotationSetValue(isoa) and
+    msoam.getNumberOfParameters() = 1 and
+    not ma.getMethod().getAParameter().getAnAnnotation().getType() instanceof TypeParam and
+    res.matches("%${%}") and
+    not res = "${" + getAnMybatisConfigurationVariableKey() + "}" and
```

Comment:
Strengthen this to use `regexpMatch` and check that there exists an X in context `${X}` that doesn't match `getAnMybatisConfigurationVariableKey()`. Otherwise we'll flag `${configKey1}${configKey2}`, or `"${configKey1}"` for example.

Reply:
Also looks like this logic is duplicated across different cases -- factor it out.

Reply:
`${getAnMybatisConfigurationVariableKey()}` is universal. I have it in the `isMybatisXmlOrAnnotationSqlInjection` predicate.